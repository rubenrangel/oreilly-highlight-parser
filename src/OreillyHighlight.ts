/**
 * A highlight from the O'Reilly Learning Platform.
 */
interface OreillyHighlight {
  /**
   * The title of the book.
   */
  bookTitle: string

  /**
   * The title of the chapter.
   */
  chapterTitle: string

  /**
   * The date the highlight was created.
   */
  dateOfHighlight: string

  /**
   * The URL of the book.
   */
  bookURL: string

  /**
   * The URL of the chapter.
   */
  chapterURL: string

  /**
   * The URL of the annotation.
   */
  annotationURL: string

  /**
   * The highlighted text.
   */
  highlight: string

  /**
   * The color of the highlight.
   */
  color: string

  /**
   * A personal note associated with the highlight.
   */
  personalNote: string
}
