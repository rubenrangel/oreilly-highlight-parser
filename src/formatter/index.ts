export { Formatter } from "./Formatter";
export { JsonFormatter } from "./JsonFormatter";
export { MarkdownFormatter } from "./MarkdownFormatter";
