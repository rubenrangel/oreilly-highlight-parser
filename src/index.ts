export { type OreillyHighlight } from "./OreillyHighlight";

export { type Formatter } from "./formatter/Formatter";
export { JsonFormatter } from "./formatter/JsonFormatter";
export { MarkdownFormatter } from "./formatter/MarkdownFormatter";

export { CsvCallbackParser } from "./parsers/CsvCallbackParser";
export { CsvStreamParser } from "./parsers/CsvStreamParser";
export { CsvSyncParser } from "./parsers/CsvSyncParser";
