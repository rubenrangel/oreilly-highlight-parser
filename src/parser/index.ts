export {
  CsvSyncParser,
  CsvCallbackParser,
  CsvStreamParser
} from "./CsvParser";
