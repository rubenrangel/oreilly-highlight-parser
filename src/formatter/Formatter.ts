import { OreillyHighlight } from "../OreillyHighlight";

/**
 * A formatter to transform highlights into a specific output.
 */
export interface Formatter {
  /**
   * Output the highlight into a formatted string.
   */
  transform(record: OreillyHighlight): string;
}
