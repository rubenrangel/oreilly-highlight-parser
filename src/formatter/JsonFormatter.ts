import { OreillyHighlight } from "../OreillyHighlight";
import { Formatter } from "./Formatter";

/**
 * Formatter to transform a highlight into a JSON string.
 */
export class JsonFormatter implements Formatter {
  transform(record: OreillyHighlight): string {
    return JSON.stringify({
      ...record,
      dateOfHighlight: record.dateOfHighlight.toString(),
    });
  }
}
