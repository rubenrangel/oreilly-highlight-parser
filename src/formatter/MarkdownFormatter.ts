import { Formatter } from "./Formatter";
import { OreillyHighlight } from "../OreillyHighlight";

/**
 * Formatter to transform a highlight into a Markdown string.
 *
 * It formats the highlight as a blockquote with the highlight text, followed by the chapter title and book title.
 */
export class MarkdownFormatter implements Formatter {
  transform(record: OreillyHighlight): string {
    const { bookTitle, annotationURL, highlight } = record;
    let { chapterTitle } = record;

    // if the chapter title starts with a number and a period, escape the period
    // so it does not act like an ordered list item in Markdown.
    if (/^\d+\./.test(chapterTitle)) {
      chapterTitle = chapterTitle.replace(".", "\\.");
    }

    return `> ${highlight}\n>\n> \\- ${chapterTitle}, [${bookTitle}](${annotationURL})`;
  }
}
