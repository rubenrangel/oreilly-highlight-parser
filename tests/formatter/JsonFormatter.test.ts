import { describe, expect, test } from "vitest";
import { OreillyHighlight } from "../../src/OreillyHighlight";
import { JsonFormatter } from "../../src/formatter/JsonFormatter";
import { Temporal } from "@js-temporal/polyfill";
import PlainDate = Temporal.PlainDate;

describe("JsonFormatter", function () {
  describe("transform", function () {
    test("it should return a JSON string", function () {
      const highlight: OreillyHighlight = {
        bookTitle: "Essential Programming",
        chapterTitle: "Chapter 1: Logic",
        dateOfHighlight: new PlainDate(2021, 9, 19),
        bookURL: "https://exmaple.com/book",
        chapterURL: "https://exmaple.com/book/chapter1",
        annotationURL: "https://exmaple.com/book/chapter1#section",
        highlight: "Using logic.",
        color: "YELLOW",
        personalNote: "This is a personal note.",
      };
      const formatter = new JsonFormatter();

      const output = formatter.transform(highlight);

      expect(output).toMatchSnapshot();
    });
  });
});
