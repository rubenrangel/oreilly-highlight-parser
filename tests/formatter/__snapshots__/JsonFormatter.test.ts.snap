// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`JsonFormatter > transform > it should return a JSON string 1`] = `"{"bookTitle":"Essential Programming","chapterTitle":"Chapter 1: Logic","dateOfHighlight":"2021-09-19","bookURL":"https://exmaple.com/book","chapterURL":"https://exmaple.com/book/chapter1","annotationURL":"https://exmaple.com/book/chapter1#section","highlight":"Using logic.","color":"YELLOW","personalNote":"This is a personal note."}"`;
